ruby_event_store/spec/link_by_metadata_spec.rb

```diff
@@ -0,0 +1,121 @@
+require 'spec_helper'
+
+module RubyEventStore
+  RSpec.describe LinkByMetadata do
+
+    let(:event_store) do
+      RubyEventStore::Client.new(repository: InMemoryRepository.new)
+    end
+
+    specify 'links to stream based on selected metadata' do
+      event_store.subscribe_to_all_events(LinkByMetadata.new(event_store: event_store, key: :string))
+      event_store.subscribe_to_all_events(LinkByMetadata.new(event_store: event_store, key: :float))
+      event_store.subscribe_to_all_events(LinkByMetadata.new(event_store: event_store, key: :int))
+      event_store.subscribe_to_all_events(LinkByMetadata.new(event_store: event_store, key: :missing))
+
+      event_store.publish(ev = OrderCreated.new(metadata:{
+        string: "city",
+        float: 1.5,
+        int: 2,
+      }))
+
+      expect(event_store.read.stream("$by_string_city").each.to_a).to eq([ev])
+      expect(event_store.read.stream("$by_float_1.5").each.to_a).to   eq([ev])
+      expect(event_store.read.stream("$by_int_2").each.to_a).to       eq([ev])
+
+      expect(event_store.read.stream("$by_missing").each.to_a).to     eq([])
+      expect(event_store.read.stream("$by_missing_").each.to_a).to    eq([])
+      expect(event_store.read.stream("$by_missing_nil").each.to_a).to eq([])
+    end
+
+    specify 'links to stream based on selected metadata (proto)' do
+      event_store = RubyEventStore::Client.new(
+        mapper: RubyEventStore::Mappers::Protobuf.new,
+        repository: InMemoryRepository.new
+      )
+      event_store.subscribe_to_all_events(LinkByMetadata.new(event_store: event_store, key: :city))
+      ev = RubyEventStore::Proto.new(
+        data: ResTesting::OrderCreated.new(
+          customer_id: 123,
+          order_id: "K3THNX9",
+        ),
+        metadata: { city: "Chicago" }
```

Comment:
I would prefer some mote "technical" example here to not encourage users to put business data into metadata.

Reply:
I see your point. However I considered this to be a technical metadata, the location of an user responsible for this event :) So you can have a look what are the people from _chicago_ doing to your system. Does that make sense?

Reply:
Chicago looks like "business" description of location. IP address would be perfectly fine here :)